From a6bcd45d3efd0d310012716bd8e241c8a085ef70 Mon Sep 17 00:00:00 2001
From: Kenneth Anthony Giusti <kgiusti@apache.org>
Date: Tue, 8 Jun 2010 16:42:35 +0000
Subject: [PATCH 05/11] Bug 595914 - QMF python console will occasionally crash if a get request times out.

QMF: ignore errors when servicing messages over a closed session.

git-svn-id: https://svn.apache.org/repos/asf/qpid/trunk@952731 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit 99811b3ed897871fb61a59712083e96a695ad04a)
---
 qpid/extras/qmf/src/py/qmf/console.py |   36 +++++++++++++++++++++++---------
 1 files changed, 26 insertions(+), 10 deletions(-)

diff --git a/qpid/extras/qmf/src/py/qmf/console.py b/qpid/extras/qmf/src/py/qmf/console.py
index d6b0ef0..6cfcdc9 100644
--- a/qpid/extras/qmf/src/py/qmf/console.py
+++ b/qpid/extras/qmf/src/py/qmf/console.py
@@ -2390,8 +2390,12 @@ class Broker:
       elif agent:
         agent._handleQmfV1Message(opcode, seq, mp, ah, codec)
 
-    self.amqpSession.receiver._completed.add(msg.id)
-    self.amqpSession.channel.session_completed(self.amqpSession.receiver._completed)
+    # ignore failures as the session may be shutting down...
+    try:
+      self.amqpSession.receiver._completed.add(msg.id)
+      self.amqpSession.channel.session_completed(self.amqpSession.receiver._completed)
+    except:
+      pass
 
   def _v2Cb(self, msg):
     """
@@ -2404,13 +2408,21 @@ class Broker:
     if 'qmf.opcode' in ah:
       opcode = ah['qmf.opcode']
       if mp.content_type == "amqp/list":
-        content = codec.read_list()
-        if not content:
-          content = []
+        try:
+          content = codec.read_list()
+          if not content:
+            content = []
+        except:
+          # malformed list - ignore
+          content = None
       elif mp.content_type == "amqp/map":
-        content = codec.read_map()
-        if not content:
-          content = {}
+        try:
+          content = codec.read_map()
+          if not content:
+            content = {}
+        except:
+          # malformed map - ignore
+          content = None
       else:
         content = None
 
@@ -2433,8 +2445,12 @@ class Broker:
             agent = self.agents[agent_addr]
             agent._handleQmfV2Message(opcode, mp, ah, content)
 
-    self.amqpSession.receiver._completed.add(msg.id)
-    self.amqpSession.channel.session_completed(self.amqpSession.receiver._completed)
+    # ignore failures as the session may be shutting down...
+    try:
+      self.amqpSession.receiver._completed.add(msg.id)
+      self.amqpSession.channel.session_completed(self.amqpSession.receiver._completed)
+    except:
+      pass
 
   def _exceptionCb(self, data):
     self.connected = False
-- 
1.7.0.1

